Step2 - 문자열 계산기

keiraJ opened the pull request:
코틀린의 여러 함수를 이용해 구현해보려고 했는데 효율적인 사용이었는지는 잘 모르겠네요 ^^;
부호와 숫자를 분리해서 validation check 를 하는 부분에서 시간이 좀 오래 걸렸습니다. (정규표현식은 여전히 어렵..ㅠ)
그리고 ktlintTestSourceSetCheck 가 자꾸 실패해서 좀 찾아보고있는데.. 흠 시간이 길어질수도 있으니 일단 리뷰를 올려보았습니다.
개선점을 찾는대로 푸시하겠습니다.
감사합니다 :)

(+) ktlinkt 문제 해결하여 푸시완료했습니다. (2022.05.07 12:25)
(+) rebase 하여 풀리퀘스트 재생성했습니다. (2022.05.08 11:25)

Commit:
step 2. 문자열 계산기

Commit:
step 2. 문자열 계산기
ktlint 문제 해결

Commit:
step 2. 문자열 계산기
main 함수 작성

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,74 @@
+import java.util.regex.Pattern
+import kotlin.jvm.Throws
+
+open class Calculator(splitedStrings: String) {
```
계산기가 단 하나의 수식만을 처리할 수 있네요.
현실의 계산기처럼 여러 개의 수식을 처리할 수 있도록 만들어보세요.

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,74 @@
+        return resultValue
+    }
+
+    private fun apply(digit1: Int, digit2: Int, operator: String): Int {
+        return when (operator) {
+            "+" -> plus(digit1, digit2)
+            "-" -> substract(digit1, digit2)
+            "*" -> multiply(digit1, digit2)
+            "/" -> divide(digit1, digit2)
+            else -> 0
+        }
+    }
```
Operator라는 별도의 클래스를 선언하고 관련된 로직들을 분리해보면 어떨까요?

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    fun calculate(): Int {
+        var resultValue = splitedValues.get(0).toInt()
```
splitedValues라는 String을 활용하는 대신 Expression이라는 클래스를 만들고 관련된 로직을 분리해보세요.

galcyurio commented on src/test/kotlin/CalculatorTest.kt:
```diff
@@ -0,0 +1,41 @@
+
+class CalculatorTest {
+
+    @ParameterizedTest
```
parameterized test 활용 👍

galcyurio commented on src/test/kotlin/CalculatorTest.kt:
```diff
@@ -0,0 +1,41 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = ["", "  "])
+    fun ShouldThrowIllegalExceptionForEmptyStrings(source: String) {
```
테스트 코드에서는 가독성이 중요합니다.
같이 일하는 동료중에 외국인이 있는게 아니라면 한글로 작성하는게 더 가독성이 좋지 않을까요?

galcyurio reviewed (changes requested):
문자열 계산기 미션 잘 진행해주셨네요 👍 
몇 가지 피드백을 남겨두었으니 확인 후 다시 요청해주세요.

Commit:
step 2. 문자열 계산기
리뷰 반영. expression 분리

Commit:
step 2. 문자열 계산기
리뷰 반영. operator 분리

Commit:
step 2. 문자열 계산기
리뷰 반영. 테스트 코드 함수명 한글로 변경

Commit:
step 2. 문자열 계산기
리뷰 반영. 여러 개의 수식을 처리할 수 있도록 변경

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -1,26 +1,28 @@
 package study
```
모든 커밋의 첫 번째 줄을 `step 2. 문자열 계산기`로 입력하신 뒤 다음 줄에 무엇을 했는지 적어주셨는데 아래 글을 참고해서 더 좋은 커밋메시지를 작성해보세요.

https://google.github.io/eng-practices/review/developer/cl-descriptions.html#firstline

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -1,26 +1,28 @@
 
-open class Calculator(splitedStrings: String) {
-    var splitedValues = listOf<String>()
+open class Calculator(inputString: String) {
```
현재는 Calculator를 상속받는 클래스가 없으므로 open 해둘 필요가 없습니다. 
꼭 필요한 경우에만 open 해주세요.

galcyurio commented on src/test/kotlin/CalculatorTest.kt:
```diff
@@ -0,0 +1,66 @@
+import org.junit.jupiter.params.provider.ValueSource
+import study.Calculator
+
+class CalculatorTest {
```
작성해주신 테스트들 중 몇 가지 케이스가 실패하고 있습니다.

galcyurio commented on src/main/kotlin/Operator.kt:
```diff
@@ -0,0 +1,28 @@
+    Plus("+") {
+        override fun apply(x: Int, y: Int): Int = x + y
+    },
+    Substract("-") {
```
`subtract`를 잘못 치신 것 같아요!

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+
+class Expression(inputStr: List<String>) {
+
+    open var operators = listOf<String>()
+    open var digits = listOf<String>()
```
불필요한 open 선언입니다.

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+        digits = inputStr.extracts(validDigitPattern, true)
+
+        val blankCount = inputStr.stream().filter { s -> s.isBlank() }.count().toInt()
+        if (blankCount > 0 || operators.size == 0 || digits.size == 0 ||
```
Intellij를 활용하다보면 가끔씩 내가 작성한 코드에 경고가 뜨는 경우가 있는데 그럴 때 해당 경고를 Quick fix 기능을 통해서 더 좋은 코드로 변경할 수 있습니다.


```kotlin
if (blankCount > 0 || operators.isEmpty() || digits.isEmpty() ||
```

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+fun List<String>.extracts(patterns: Pattern, match: Boolean): List<String> {
+    val list = mutableListOf<String>()
+    this.stream().forEach { a ->
+        if (patterns.matcher(a).find() == match) {
+            list.add(a)
+        }
+    }
+
+    return list
+}
+
+@Throws(IllegalArgumentException::class)
+fun List<String>.validate(patterns: Pattern): Boolean {
+    return this.stream().allMatch { a -> patterns.matcher(a).find() }
+}
```
불필요한 확장함수 사용입니다.
이 함수들은 Expression 클래스 내부에 `private fun`으로 선언해주세요.

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    companion object {
+        var DIGITS_REGEX = "^[0-9]*$"
+        var OPERATORS_REGEX = "[*+-/]"
+        val validDigitPattern: Pattern = Pattern.compile(DIGITS_REGEX)
+        val validOperatorPattern: Pattern = Pattern.compile(OPERATORS_REGEX)
```
불필요한 var, public 선언입니다.
모두 `private val`로 선언해주세요.

-----
필드를 선언할 때 항상 `private val`로 선언하고 꼭 필요할 때만 풀어주는 습관을 들여보면 좋겠어요.

galcyurio replied:
const가 아니더라도 해당 객체가 불변이라면 UPPER_CASE로 지어주세요.

> Constants are `val` properties with no custom `get` function, whose contents are deeply immutable, and whose functions have no detectable side-effects. This includes immutable types and immutable collections of immutable types as well as scalars and string if marked as `const`. If any of an instance’s observable state can change, it is not a constant. Merely intending to never mutate the object is not enough.
> https://developer.android.com/kotlin/style-guide#constant_names

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+fun List<String>.extracts(patterns: Pattern, match: Boolean): List<String> {
```
함수 시그니쳐로 MutableList를 사용하지 않고 List 사용 👍

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+
+fun List<String>.extracts(patterns: Pattern, match: Boolean): List<String> {
+    val list = mutableListOf<String>()
+    this.stream().forEach { a ->
```
코틀린에서는 자바의 Stream을 사용할 필요없이 코틀린의 Sequence를 사용하시면 됩니다.

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+}
+
+fun List<String>.extracts(patterns: Pattern, match: Boolean): List<String> {
+    val list = mutableListOf<String>()
+    this.stream().forEach { a ->
+        if (patterns.matcher(a).find() == match) {
+            list.add(a)
+        }
+    }
+
+    return list
+}
```
코틀린의 List, Set과 같은 컬렉션은 특이하게도 자바의 Stream에서 제공하는 map(), filter()와 같은 함수를 바로 호출할 수 있습니다.

```kotlin
fun List<String>.extracts(patterns: Pattern, match: Boolean): List<String> {
    return filter { patterns.matcher(it).find() == match }
}
```

차이점은 컬렉션에서 호출하면 바로 호출되고 Sequence에서 호출하면 lazy하게 처리됩니다.

galcyurio reviewed (changes requested):
피드백 잘 반영해주셨습니다.
추가로 몇 가지 코멘트 남겨두었으니 확인해주세요!
궁금한 부분 있으면 언제든지 물어보셔도 좋습니다!

Commit:
step 2. 문자열 계산기
리뷰 반영>
1) 불필요한 var와 open 사용 제거 및 private화
2) substract -> subtract 오타 수정
3) 붋변하는 const,, 대문자로 선언하도록 수정

Commit:
step 2. 문자열 계산기
리뷰 반영>
1) sequece를 활용하도록 코드 수정
2) 테스트 코드 실패 보정

keiraJ commented:
시퀀스에 대해서 구글링을 통해 이점을 확인했는데요, 아직 익숙하지 않아서 잘 적용한 것인지 모르겠습니다 ㅎㅎ;
이번 기회에 책을 좀 더 읽어보고 많이 익숙해져봐야겠네요!
감사합니다! ☺️

Commit:
step 2. 문자열 계산기
리뷰 반영>
1) quick fix 를 이용해 수식 정리하기

Commit:
step 2. 문자열 계산기
리뷰 반영>
1) stream 제거
2) unused chain 제거

galcyurio commented on src/main/kotlin/Calculator.kt:
```diff
@@ -3,7 +3,7 @@ package study
 import Expression
```
앗! 커밋 메시지에 대한 피드백을 오해하신 것 같습니다.
공유드린 구글 코드리뷰 가이드를 요약하자면 아래와 같습니다.

- 첫 번째 줄에는 해당 커밋에서 발생한 변경점들을 요약하여 작성한다.
- 두 번째 줄은 비우고
- 세 번째 줄부터는 왜 이렇게 변경해야 했는지, 왜 이게 최선인지에 대해 자세하게 적는다.

-----
개인적으로는 대부분의 경우 첫 번째 줄만 작성하고 자세한 설명이 필요한 경우에만 길게 작성하는 편입니다.

galcyurio commented on src/main/kotlin/Expression.kt:
```diff
@@ -27,17 +27,14 @@ class Expression(inputStr: List<String>) {
 @Throws(IllegalArgumentException::class)
 fun List<String>.validate(patterns: Pattern): Boolean {
-    return this.stream().allMatch { a -> patterns.matcher(a).find() }
+    val seq = this.asSequence()
+    return seq.all {
+        a ->
+        patterns.matcher(a).find()
+    }
```
Sequence 잘 활용해주셨네요 👍 

-----

Sequence는 큰 컬렉션에서 많은 작업을 수행할 때 유리합니다.
그 이외에는 굳이 Sequence를 쓸 필요없이 컬렉션에서 처리하셔도 좋습니다.
https://blog.kotlin-academy.com/effective-kotlin-use-sequence-for-bigger-collections-with-more-than-one-processing-step-649a15bb4bf

```kotlin
fun main() {
    val intRange = (1..10_000_000)
    measureTimeMillis {
        intRange.toList()
            .map { it * it }
            .filter { it < 1_000_000 }
            .map { it - 2 }
            .filter { it < 1000 }
            .map { it * 10 }
            .first()
    }.also { println("List: $it") } // List: 6807
    measureTimeMillis {
        intRange.asSequence()
            .map { it * it }
            .filter { it < 1_000_000 }
            .map { it - 2 }
            .filter { it < 1000 }
            .map { it * 10 }
            .first()
    }.also { println("Sequence: $it") } // Sequence: 8
}
```

galcyurio reviewed (approved):
계산기 미션 고생하셨습니다 👏 👏 
추가로 알아두시면 좋을만한 내용이 있어서 코멘트 남겨두었어요!
자동차 경주 미션도 힘내세요!
